src/pipeline.rs

```diff
@@ -382,3 +430,98 @@ pub(crate) fn prepare_view_targets(
+            speed: 5.0,
+        }
+    }
+}
+
+impl ExtractResource for PointCloudPlaybackControl {
+    type Source = Self;
+
+    /// Defines how the resource is transferred into the "render world".
+    fn extract_resource(source: &Self::Source) -> Self {
+        source.clone()
+    }
+}
+
+pub fn prepare_animated_assets(
+    queue: Res<RenderQueue>,
+    mut assets: ResMut<RenderAssets<PointCloudAsset>>,
+    playback: ResMut<PointCloudPlaybackControl>,
+    time: Res<Time>,
+) {
+    if !playback.playing {
+        return;
+    }
+    for (_handle, asset) in assets.iter_mut() {
+        if let Some(animation_buffer) = asset.animation_buffer.as_mut() {
+            let mut view = vec![0.0; asset.num_points as usize * 3];
+
+            match asset.frames.as_ref().unwrap() {
+                opd_parser::Frames::I8(frames) => {
+                    let current_frame = &frames[asset.current_animation_frame];
+                    asset.animation_time += time.delta_seconds() * playback.speed;
+
+                    if current_frame.time / 1000.0 > asset.animation_time {
+                        continue;
+                    }
+
+                    asset.current_animation_frame += 1;
+                    if asset.current_animation_frame >= frames.len() {
+                        asset.current_animation_frame = 0;
+                        asset.animation_time = 0.0;
+                    }
+
+                    let scale: [f32; 3] = asset.animation_scale.into();
+                    let mut iter = frames[asset.current_animation_frame]
+                        .data
+                        .iter()
+                        .enumerate();
+                    loop {
+                        let arr = [
+                            match iter.next() {
+                                Some(a) => (a.0, *a.1),
+                                None => break,
+                            },
+                            match iter.next() {
+                                Some(a) => (a.0, *a.1),
+                                None => break,
+                            },
+                            match iter.next() {
+                                Some(a) => (a.0, *a.1),
+                                None => break,
+                            },
+                        ];
+                        for j in 0..3 {
+                            let (i, value) = arr[j];
+                            let value: f32 = value as f32 / i8::MAX as f32;
+
+                            let scale = scale[j];
+                            view[i] = value * scale;
+                        }
+                    }
+
+                    frames.len()
+                }
+                _ => todo!(),
+            };
+
+            queue.write_buffer(animation_buffer, 0, unsafe {
+                std::slice::from_raw_parts(
+                    view.as_ptr() as *const u8,
+                    std::mem::size_of_val(view.as_slice()),
+                )
```

Comment:
Check out the `bytemuck` crate for a safer way to do this.

https://docs.rs/bytemuck/latest/bytemuck/fn.cast_slice.html

Reply:
Longer term we'd be using wgpu's `write_buffer_with` method to reduce copy (there's currently a synchronization bug with that method), so let's not introduce an additional dependency for this one thing.

Reply:
`bytemuck` is pretty small. It's already used all over the place in Foresight code.